numpy/random/mtrand.pyx

```diff
@@ -3510,10 +3510,11 @@ cdef class RandomState:
         ...    print(i, "wells drilled, probability of one success =", probability)
 
         """
-        out = disc(&legacy_negative_binomial, &self._aug_state, size, self.lock, 2, 0,
+        out = disc(&legacy_negative_binomial, &self._aug_state, size, self.lock, 2, 0, 0,
```

Comment:
Changes to mtrand aren't allowed in general if they would change the stream.  All that could be done here would be to protect against values that lead to infinite loops.

Reply:
Could you clarify what do you mean with changing the stream? Is changing the signature of ```disc``` not allowed? 

Otherwise, note that I didn't change the output of any of the functions in mtrand at all (not even to protect against the infinite loops, calling ```np.random.negative_binomial(2**64, 0.1)``` would still result in an infinite loop, the checks only apply to the calls to ```default_rng().negative_binomial()```), I just updated the function calls since the new ```disc``` signature has an extra parameter, which is set to an ignored value for all functions in mtrand.

Reply:
If a value slightly above lam max in your code would eventually return, then you have now changed the stream.  I haven't reasoned carefully about this but as a general rule mtrand only changes for trivial things like doc string fixes and to keep to compiling. 

See https://numpy.org/neps/nep-0019-rng-policy.html

Reply:
Understood your point about the stream. However, upon reading NEP-19 on your link, it's still unclear to me if this applies to the new PRNGs too (i.e. are changes to ```np.random.default_rng().negative_binomial```'s stream allowed?). 

Calls to ```np.random.negative_binomial``` (the one in mtrand) in this PR do **not** raise an error even if they would result in a value of lam too large to ever return, the stream is not affected for any function in mtrand. Only the ```negative_binomial``` distribution that is created when calling a ```Generator``` method is affected.

Reply:
Changes and improvements to `Generator` are definitely allowed.  Changes like this that prevent infinite loops, which are really bugs, are definitely in bounds. I think it would help if you removed any changes to mtrand here so it would be easier to think about the changes to the checks independently from the NEP issues.